cmd/clusterctl/client/cluster/internal/dryrun/client.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dryrun
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/scheme"
+)
+
+var (
+	localScheme = scheme.Scheme
+)
+
+type changeTrackerID struct {
+	gvk schema.GroupVersionKind
+	key client.ObjectKey
+}
+
+// changeTracker is used to track the list of operations that are performed
+// and the objects that the operations are performed on.
+// Nb. Objects that are both created and modified are tracked only under created.
+//
+// While changeTracker tracks the operations on objects using unique object identifiers
+// ChangeSummary reports all the operations and the final state of objects. ChangeSummary
+// is calculated using changeTracker and fake client.
+type changeTracker struct {
+	created map[changeTrackerID]bool
+	patched map[changeTrackerID]client.Object
+	deleted map[changeTrackerID]bool
+}
+
+// Client implements a dry run Client, that is a fake.Client that logs write operations.
+type Client struct {
+	fakeClient client.Client
+	apiReader  client.Reader
+
+	changeTracker *changeTracker
+}
+
+// PatchSummary defined the patch observed on an object.
+type PatchSummary struct {
+	Before *unstructured.Unstructured
+	After  *unstructured.Unstructured
+}
+
+// ChangeSummary defines all the changes detected by the Dryrun execution.
+// Nb. Objects that are both created and modified are tracked only under created.
```

Comment:
let's makes this generic, by saying that we will report a single operation for each object, flattening operations to highlight the diff before/after (vs highlighting all the intermediate steps). e.g. create + upgrade --> create